subsys/net/lib/lwm2m/lwm2m_obj_firmware.c

```diff
@@ -319,27 +345,32 @@ static struct lwm2m_engine_obj_inst *firmware_create(uint16_t obj_inst_id)
 {
 	int i = 0, j = 0;
 
-	init_res_instance(res_inst, ARRAY_SIZE(res_inst));
+	init_res_instance(res_inst[obj_inst_id], ARRAY_SIZE(res_inst[obj_inst_id]));
 
 	/* initialize instance resource data */
-	INIT_OBJ_RES_OPT(FIRMWARE_PACKAGE_ID, res, i, res_inst, j, 1, false,
-			 true, NULL, NULL, NULL, package_write_cb, NULL);
-	INIT_OBJ_RES(FIRMWARE_PACKAGE_URI_ID, res, i, res_inst, j, 1, false,
-		     true, package_uri, PACKAGE_URI_LEN,
-		     NULL, NULL, NULL, package_uri_write_cb, NULL);
-	INIT_OBJ_RES_EXECUTE(FIRMWARE_UPDATE_ID, res, i, firmware_update_cb);
-	INIT_OBJ_RES_DATA(FIRMWARE_STATE_ID, res, i, res_inst, j,
-			  &update_state, sizeof(update_state));
-	INIT_OBJ_RES_DATA(FIRMWARE_UPDATE_RESULT_ID, res, i, res_inst, j,
-			  &update_result, sizeof(update_result));
-	INIT_OBJ_RES_DATA(FIRMWARE_UPDATE_DELIV_METHOD_ID, res, i, res_inst, j,
-			  &delivery_method, sizeof(delivery_method));
-
-	inst.resources = res;
-	inst.resource_count = i;
+	INIT_OBJ_RES_OPT(FIRMWARE_PACKAGE_ID, res[obj_inst_id], i, res_inst[obj_inst_id], j, 1,
+			 false, true, NULL, NULL, NULL, package_write_cb, NULL);
+	INIT_OBJ_RES(FIRMWARE_PACKAGE_URI_ID, res[obj_inst_id], i, res_inst[obj_inst_id], j, 1,
+		     false, true, package_uri[obj_inst_id], PACKAGE_URI_LEN, NULL, NULL, NULL,
+		     package_uri_write_cb, NULL);
+	INIT_OBJ_RES_EXECUTE(FIRMWARE_UPDATE_ID, res[obj_inst_id], i, firmware_update_cb);
+	INIT_OBJ_RES_DATA(FIRMWARE_STATE_ID, res[obj_inst_id], i, res_inst[obj_inst_id], j,
+			  &(update_state[obj_inst_id]), sizeof(update_state[obj_inst_id]));
+	INIT_OBJ_RES_DATA(FIRMWARE_UPDATE_RESULT_ID, res[obj_inst_id], i, res_inst[obj_inst_id], j,
+			  &(update_result[obj_inst_id]), sizeof(update_result[obj_inst_id]));
+	INIT_OBJ_RES_OPT(FIRMWARE_PACKAGE_NAME_ID, res[obj_inst_id], i, res_inst[obj_inst_id], j, 1,
+			 false, true, NULL, NULL, NULL, NULL, NULL);
+	INIT_OBJ_RES_OPT(FIRMWARE_PACKAGE_VERSION_ID, res[obj_inst_id], i, res_inst[obj_inst_id], j,
```

Comment:
It think those Optional unbuffered resources could use INIT_OBJ_RES_OPTDATA() but that is calling INIT_OBJ_RES_OPT() with NULL like your code.  Correct me If I'm wrong.

Reply:
Yes they could, behaviour would be exactly the same though.